### PR TITLE
feat: customer.io webhook for marketing cta

### PR DESCRIPTION
**Pull request by omBratteng**
Adds webhook endpoint for use with customer.io and assigning users to marketing CTA.
The POST requests have to have a valid HMAC signature

AS-328

### CODE REVIEW

File: src/routes/webhooks.ts [hunk 1]
```diff
@@ -109,4 +141,36 @@ export default async function (fastify: FastifyInstance): Promise<void> {
...
+      const con = await createOrGetConnection();
+      await con.getRepository(UserMarketingCta).upsert(
+        {
+          userId: userId,
+          marketingCtaId: marketingCtaId,
+        },
+        ['userId', 'marketingCtaId'],
+      );
+
+      await getMarketingCta(con, logger, userId);
```
Comment (omBratteng) [inline src/routes/webhooks.ts:171]: It'll throw a 500 error of either of userId or marketingCtaId doesn't exist.
Should probably wrap it in a try...catch so that we can return a different status

File: src/schema/users.ts [hunk 2]
```diff
@@ -773,31 +800,8 @@ export const getMarketingCta = async (
...
-    const redisValue = userMarketingCta
-      ? JSON.stringify(marketingCta)
-      : RedisMagicValues.SLEEPING;
-
-    await setRedisObjectWithExpiry(redisKey, redisValue, ONE_WEEK_IN_SECONDS);
-  }
-
-  return marketingCta;
+  const marketingCta: MarketingCta | null = JSON.parse(rawRedisValue || null);
+  return marketingCta || cachePrefillMarketingCta(con, userId, redisKey);
```
Comment (omBratteng) [inline src/schema/users.ts:804]: Didn't need this change, should I keep it or revert it?

File: src/routes/webhooks.ts (hunk 1 above)
Comment (capJavert) [reply src/routes/webhooks.ts:171]: Yeah, webhoook providers often disable webhooks that return 2xx status code repediately so I would prefer we catch, log error and return 2xx always.

File: src/schema/users.ts (hunk 2 above)
Comment (capJavert) [reply src/schema/users.ts:804]: If we don't need it lets remove. Less code is better.

File: __tests__/webhooks.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        .post('/webhooks/customerio/marketing_cta')
+        .set('x-cio-timestamp', timestamp.toString())
+        .set('x-cio-signature', hash)
+        .send(payload)
+        .expect(200);
+
+      expect(body.success).toEqual(true);
+    });
+
+    it('should return 200 when signature is valid (dynamic)', async () => {
```
Comment (capJavert) [inline __tests__/webhooks.ts:104]: what does `dynamic` mean in test name?

File: src/routes/webhooks.ts [hunk 4]
```diff
@@ -45,6 +51,32 @@ const verifyRequest = (publicKey: string, req: FastifyRequest): boolean => {
...
+  hmac.update(req.rawBody);
+
+  const hash = hmac.digest();
+
+  if (!timingSafeEqual(hash, Buffer.from(signature, 'hex'))) {
+    logger.debug("Signature didn't match");
+    return false;
+  }
+
+  logger.debug('Signature matched!');
```
Comment (capJavert) [inline src/routes/webhooks.ts:76]: I know it is just debug logging, but probably something like "CIO signature matched"? Same above.

File: src/workers/cdc/primary.ts [hunk 5]
```diff
@@ -878,9 +854,6 @@ const worker: Worker = {
         case getTableName(con, MarketingCta):
           await onMarketingCtaChange(con, data);
           break;
-        case getTableName(con, UserMarketingCta):
-          await onUserMarketingCtaChange(data);
-          break;
```
Comment (capJavert) [inline src/workers/cdc/primary.ts:883]: Nice!

File: src/routes/webhooks.ts (hunk 1 above)
Comment (capJavert) [reply src/routes/webhooks.ts:171]: We can return `success: false` in that case as well.

**Review by capJavert — COMMENTED** [review]
Looks good, looking forward to implementation. Few comments.

File: __tests__/webhooks.ts (hunk 3 above)
Comment (omBratteng) [reply __tests__/webhooks.ts:104]: The other one (static) has the values hardcoded. And this one uses real timestamp for hmac signature

Comment (capJavert) [reply __tests__/webhooks.ts:104]: So its the same test?

Comment (omBratteng) [reply __tests__/webhooks.ts:104]: Basically yeah

Comment (capJavert) [reply __tests__/webhooks.ts:104]: Never enough tests but interested what is the benefit of having both? Any edge case that you found?

Comment (omBratteng) [reply __tests__/webhooks.ts:104]: No edge cases, but my reasoning is if we modify the signature checker we'll also modify the signature generator in the test, so we might introduce the same error in both places. Then the static test will stand as a safeguard

Comment (capJavert) [reply __tests__/webhooks.ts:104]: Makes sense 👍

File: src/routes/webhooks.ts (hunk 4 above)
Comment (omBratteng) [reply src/routes/webhooks.ts]: Added

File: src/schema/users.ts (hunk 2 above)
Comment (omBratteng) [reply src/schema/users.ts]: Reverted it

**Review by rebelchris — APPROVED** [review]
Nice work, besides Ante's comment nothing to add.

File: src/routes/webhooks.ts (hunk 1 above)
Comment (omBratteng) [reply src/routes/webhooks.ts:171]: 👍 I will put it in a try...catch and return false and logs if it fails.

Comment (omBratteng) [reply src/routes/webhooks.ts]: I return 400 with `success:false`

Comment (omBratteng) [reply src/routes/webhooks.ts]: > We have a 4 second timeout and if we get a 408, 409, 429, or 5xx response code in that period, Customer.io will retry sending the notification several times with exponential backoff. We will retry your request up to 10 times over a period of approximately 3 hours.

File: src/workers/cdc/primary.ts [hunk 6]
```diff
@@ -771,30 +771,6 @@ const onMarketingCtaChange = async (
...
-    if (data.payload.before.readAt !== null) return;
-
-    await deleteRedisKey(
-      generateStorageKey(
-        StorageTopic.Boot,
-        StorageKey.MarketingCta,
-        data.payload.before.userId,
-      ),
-    );
-  }
```
Comment (omBratteng) [inline src/workers/cdc/primary.ts:795]: I will create a another PR for the deletion webhook

File: .infra/Pulumi.prod.yaml [hunk 7]
```diff
@@ -46,6 +46,8 @@ config:
       secure: AAABAKsjQ79BKL/Msk8ERCFq7TeKGQ6NnzRcQrOoUzQ7hFEDNdq4uxOcG3w5M5aODolHDw==
     cioSiteId:
       secure: AAABAACN2IeoykRGHprr1nYvk1snUvT733+1dp8QI7+WchlaErfJJ3X6LLPYe8BthHnpqw==
+    cioWebhookSecret:
+      secure: AAABAHkXsqVvvtA1VZ3lLrNQoV0Kh3FM31l7QxQVVwQvR4CHbLyqbtg2N+MJ5URkaRwYsMMPv+8RNnRcCiC/v33GjNBc4e57x5MpVpQC2VPUyjMornJI53n7rmeQBtsQ
```
Comment (omBratteng) [inline .infra/Pulumi.prod.yaml:50]: This is the testing workspace secret, will revert to production/daily.dev workspace once we know it actually works 🙈

File: src/routes/webhooks.ts [hunk 8]
```diff
@@ -109,4 +141,41 @@ export default async function (fastify: FastifyInstance): Promise<void> {
...
+            marketingCtaId: marketingCtaId,
+          },
+          ['userId', 'marketingCtaId'],
+        );
+
+        await getMarketingCta(con, logger, userId);
+
+        return res.send({ success: true });
+      } catch (error) {
+        logger.error(error, 'Error processing CIO webhook');
```
Comment (capJavert) [inline src/routes/webhooks.ts:176]: We usually send `logger.error({ err: error }, 'Error processing CIO webhook'');`, I think because GC then extracts from `err`?

File: src/routes/webhooks.ts (hunk 1 above)
Comment (capJavert) [reply src/routes/webhooks.ts]: Nice 👍

**Review by capJavert — APPROVED** [review]
🚀
